### PR TITLE
auto renew vault token if it is nearing expiry

### PR DESCRIPTION
**Pull request by deontpearson**
My aim was to store the TTL of the token in the service. The token will be renewed in the last hour of the tokens lifetime. Some things I needed to achieve:
- Only request 1 token for the whole service. The Class will share this token
- Do not make repeated requests to Vault to check if the token has expired.

### CODE REVIEW

File: src/main/java/com/github/jcustenborder/kafka/config/vault/KubernetesAuth.java [hunk 1]
```diff
@@ -16,60 +16,45 @@
...
+import java.nio.file.Path;
 import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class KubernetesAuth {
-  private static final Logger log = LoggerFactory.getLogger(AuthHandlers.class);
```
Comment (seapen) [inline src/main/java/com/github/jcustenborder/kafka/config/vault/KubernetesAuth.java:30]: Just confirm other classes have this empty line.

File: src/main/java/com/github/jcustenborder/kafka/config/vault/KubernetesAuth.java [hunk 2]
```diff
@@ -16,60 +16,45 @@
...
-
-      if (Strings.isNullOrEmpty(jwt)) {
-        throw new Exception(String.format("JWT token from file is invalid (%s)", jwt));
-      }
-      
-      AuthResponse authResponse = vault.auth().loginByKubernetes(role, jwt);
-      return authResponse.getAuthClientToken();
-    } catch (Exception e) {
-      throw new RuntimeException(e);
+      String jwt = getJWTFromFile(this.config.jwtPath);
```
Comment (seapen) [inline src/main/java/com/github/jcustenborder/kafka/config/vault/KubernetesAuth.java:49]: Would it not be better to try read this into memory once?

Comment (deontpearson) [reply src/main/java/com/github/jcustenborder/kafka/config/vault/KubernetesAuth.java:49]: maybe read it once when the instance is created? 

image

File: src/main/java/com/github/jcustenborder/kafka/config/vault/VaultConfigProvider.java [hunk 3]
```diff
@@ -124,7 +123,80 @@ public void configure(Map<String, ?> settings) {
...
+    LocalDateTime tokenExpiryTime = getTokenExpirationTime();
+    TOKEN_METADATA.updateAndGet(old -> new TokenMetadata(renewedToken, tokenExpiryTime));
+  }
+
+  private void ensureVaultIsConfigured() {
+    log.info("Ensuring Vault is configured");
+    if (this.vault == null) throw new RuntimeException("Vault is not configured");
+
+    try {
+      if (this.config.loginBy == VaultLoginBy.Token && !tokenIsValid()) renewToken();
```
Comment (jfoster94) [inline src/main/java/com/github/jcustenborder/kafka/config/vault/VaultConfigProvider.java:167]: Technically we could only do one `tokenIsValid` call right?

**Review by jfoster94 — APPROVED** [review]

**Review by jfoster94 — COMMENTED** [review]
This looks fine to me. Lets try the simpler approach we discussed and fallback to this if necessary.

**Review by gatanga — APPROVED** [review]

**Pull request merged**